diff -Nru a/data/gnome-packagekit.schemas.in b/data/gnome-packagekit.schemas.in
--- a/data/gnome-packagekit.schemas.in	2010-05-28 02:39:59.000000000 +1200
+++ b/data/gnome-packagekit.schemas.in	2010-07-28 12:58:20.737896631 +1200
@@ -62,6 +62,18 @@
     </schema>
 
     <schema>
+      <key>/schemas/apps/gnome-packagekit/enable_autoremove</key>
+      <applyto>/apps/gnome-packagekit/enable_autoremove</applyto>
+      <owner>gnome-packagekit</owner>
+      <type>bool</type>
+      <default>false</default>
+      <locale name="C">
+        <short>Automatically remove unused dependencies</short>
+        <long>When removing a package, also remove those dependencies that are not required by other packages</long>
+      </locale>
+    </schema>
+
+    <schema>
       <key>/schemas/apps/gnome-packagekit/show_depends</key>
       <applyto>/apps/gnome-packagekit/show_depends</applyto>
       <owner>gnome-packagekit</owner>
@@ -324,7 +336,7 @@
       <applyto>/apps/gnome-packagekit/update-icon/frequency_get_upgrades</applyto>
       <owner>gnome-packagekit</owner>
       <type>int</type>
-      <default>604800</default>
+      <default>0</default>
       <locale name="C">
         <short>How often to check for distribution upgrades</short>
         <long>How often to check for distribution upgrades. Value is in seconds.</long>
@@ -540,7 +552,7 @@
       <applyto>/apps/gnome_settings_daemon/gtk-modules/pk-gtk-module</applyto>
       <owner>gnome-packagekit</owner>
       <type>bool</type>
-      <default>true</default>
+      <default>false</default>
       <locale name="C">
         <short>GTK+ module for font installation</short>
         <long>This key determines if applications should be able to prompt for fonts</long>
diff -Nru a/data/gpk-install-file.desktop.in b/data/gpk-install-file.desktop.in
--- a/data/gpk-install-file.desktop.in	2010-05-27 22:15:01.000000000 +1200
+++ b/data/gpk-install-file.desktop.in	2010-07-28 12:58:20.741229964 +1200
@@ -10,5 +10,5 @@
 Icon=system-software-install
 StartupNotify=true
 NoDisplay=true
-MimeType=application/x-rpm;application/x-redhat-package-manager;application/x-servicepack;application/x-deb;
+MimeType=application/x-compressed-tar;application/x-xz-compressed-tar;application/x-servicepack;
 
diff -Nru a/src/gpk-application.c b/src/gpk-application.c
--- a/src/gpk-application.c	2010-05-28 02:55:37.000000000 +1200
+++ b/src/gpk-application.c	2010-08-18 18:48:37.577310099 +1200
@@ -1024,7 +1024,8 @@
 
 	/* get the requires */
 	package_ids = pk_package_ids_from_id (package_id_selected);
-	pk_client_get_requires_async (PK_CLIENT(application->priv->task), PK_FILTER_ENUM_NONE,
+	pk_client_get_requires_async (PK_CLIENT(application->priv->task),
+				      pk_bitfield_value (PK_FILTER_ENUM_NONE),
 				      package_ids, TRUE, application->priv->cancellable,
 				      (PkProgressCallback) gpk_application_progress_cb, application,
 				      (GAsyncReadyCallback) gpk_application_get_requires_cb, application);
@@ -1147,7 +1148,8 @@
 
 	/* get the depends */
 	package_ids = pk_package_ids_from_id (package_id_selected);
-	pk_client_get_depends_async (PK_CLIENT(application->priv->task), PK_FILTER_ENUM_NONE,
+	pk_client_get_depends_async (PK_CLIENT(application->priv->task),
+				     pk_bitfield_value (PK_FILTER_ENUM_NONE),
 				     package_ids, TRUE, application->priv->cancellable,
 				     (PkProgressCallback) gpk_application_progress_cb, application,
 				     (GAsyncReadyCallback) gpk_application_get_depends_cb, application);
@@ -2115,6 +2117,7 @@
 gpk_application_button_apply_cb (GtkWidget *widget, GpkApplication *application)
 {
 	gchar **package_ids = NULL;
+	gboolean autoremove;
 
 	g_return_if_fail (GPK_IS_APPLICATION (application));
 
@@ -2132,8 +2135,10 @@
 
 	} else if (application->priv->action == PK_ACTION_REMOVE) {
 
-		/* install */
-		pk_task_remove_packages_async (application->priv->task, package_ids, TRUE, FALSE, application->priv->cancellable,
+		autoremove = gconf_client_get_bool (application->priv->gconf_client, GPK_CONF_ENABLE_AUTOREMOVE, NULL);
+
+		/* remove */
+		pk_task_remove_packages_async (application->priv->task, package_ids, TRUE, autoremove, application->priv->cancellable,
 					       (PkProgressCallback) gpk_application_progress_cb, application,
 					       (GAsyncReadyCallback) gpk_application_remove_packages_cb, application);
 
@@ -4175,7 +4180,9 @@
 			  G_CALLBACK (gpk_application_groups_treeview_changed_cb), application);
 
 	/* get repos, so we can show the full name in the software source box */
-	pk_client_get_repo_list_async (PK_CLIENT(application->priv->task), PK_FILTER_ENUM_NONE, application->priv->cancellable,
+	pk_client_get_repo_list_async (PK_CLIENT(application->priv->task),
+				       pk_bitfield_value (PK_FILTER_ENUM_NONE),
+				       application->priv->cancellable,
 				       (PkProgressCallback) gpk_application_progress_cb, application,
 				       (GAsyncReadyCallback) gpk_application_get_repo_list_cb, application);
 
diff -Nru a/src/gpk-check-update.c b/src/gpk-check-update.c
--- a/src/gpk-check-update.c	2010-05-28 02:39:25.000000000 +1200
+++ b/src/gpk-check-update.c	2010-07-28 13:00:19.037053800 +1200
@@ -1095,7 +1095,9 @@
 	}
 
 	/* get new update list */
-	pk_client_get_updates_async (PK_CLIENT(cupdate->priv->task), PK_FILTER_ENUM_NONE, cupdate->priv->cancellable, NULL, NULL,
+	pk_client_get_updates_async (PK_CLIENT(cupdate->priv->task),
+				     pk_bitfield_value (PK_FILTER_ENUM_NONE),
+				     cupdate->priv->cancellable, NULL, NULL,
 				     (GAsyncReadyCallback) gpk_check_update_get_updates_finished_cb, cupdate);
 out:
 	return;
diff -Nru a/src/gpk-common.h b/src/gpk-common.h
--- a/src/gpk-common.h	2010-05-28 02:39:25.000000000 +1200
+++ b/src/gpk-common.h	2010-07-28 12:58:20.744568048 +1200
@@ -37,6 +37,7 @@
 #define GPK_CONF_SHOW_COPY_CONFIRM		"/apps/gnome-packagekit/show_copy_confirm"
 #define GPK_CONF_DBUS_DEFAULT_INTERACTION	"/apps/gnome-packagekit/dbus_default_interaction"
 #define GPK_CONF_DBUS_ENFORCED_INTERACTION	"/apps/gnome-packagekit/dbus_enforced_interaction"
+#define GPK_CONF_ENABLE_AUTOREMOVE		"/apps/gnome-packagekit/enable_autoremove"
 #define GPK_CONF_ENABLE_FONT_HELPER		"/apps/gnome-packagekit/enable_font_helper"
 #define GPK_CONF_ENABLE_CODEC_HELPER		"/apps/gnome-packagekit/enable_codec_helper"
 #define GPK_CONF_ENABLE_MIME_TYPE_HELPER	"/apps/gnome-packagekit/enable_mime_type_helper"
