diff -ru a/src/gpk-application.c b/src/gpk-application.c
--- a/src/gpk-application.c	2010-01-29 00:19:06.000000000 +1300
+++ b/src/gpk-application.c	2010-02-23 01:37:45.366262919 +1300
@@ -2133,7 +2133,7 @@
 	} else if (application->priv->action == PK_ACTION_REMOVE) {
 
 		/* install */
-		pk_task_remove_packages_async (application->priv->task, package_ids, TRUE, FALSE, application->priv->cancellable,
+		pk_task_remove_packages_async (application->priv->task, package_ids, FALSE, TRUE, application->priv->cancellable,
 					       (PkProgressCallback) gpk_application_progress_cb, application,
 					       (GAsyncReadyCallback) gpk_application_remove_packages_cb, application);
 
