diff -ru a/src/gpk-application.c b/src/gpk-application.c
--- a/src/gpk-application.c	2010-02-16 06:54:32.000000000 +1300
+++ b/src/gpk-application.c	2010-03-31 15:59:30.963387825 +1300
@@ -4039,6 +4039,7 @@
 	widget = GTK_WIDGET (gtk_builder_get_object (application->priv->builder, "button_cancel"));
 	g_signal_connect (widget, "clicked",
 			  G_CALLBACK (gpk_application_cancel_cb), application);
+	gtk_widget_set_sensitive (widget, FALSE);
 	/* TRANSLATORS: tooltip on the cancel button */
 	gtk_widget_set_tooltip_text (widget, _("Cancel search"));
 
